src/sqlfluff/utils/reflow/reindent.py

```diff
@@ -641,11 +643,28 @@ def _crawl_indent_points(
 
 def _map_line_buffers(
     elements: ReflowSequenceType, allow_implicit_indents: bool = False
-) -> List[_IndentLine]:
-    """Map the existing elements, building up a list of _IndentLine."""
+) -> Tuple[List[_IndentLine], List[int]]:
+    """Map the existing elements, building up a list of _IndentLine.
+
+    Returns:
+        :obj:`tuple` of a :obj:`list` of :obj:`_IndentLine` and a
+            :obj:`list` of :obj:`int`. The first is the main output
+            and is designed to be used in assessing indents and
+            their effect through a SQL file. The latter is a list of
+            "imbalanced" indent locations, where the positive indent
+            is untaken, but it's corresponding negative indent *is*
+            taken.
+
+    """
     # First build up the buffer of lines.
     lines = []
     point_buffer = []
+    # Buffers to keep track of indents which are untaken on the way
+    # up but taken on the way down. We track them explicitly so we
+    # can force them later.
+    untaken_indent_locs = {}
```

Comment:
Maybe consider documenting what are the keys/values in these two variables, not just types but what they _mean_. (E.g. if it's just an integer, is that an index into some list, or a character position...).

Reply:
Yeah - good idea. The naming isn't super clear here.